inference-engine/tests/functional/plugin/myriad/subgraph_tests/dsr_tests_common.hpp

```diff
@@ -0,0 +1,106 @@
+        const auto inDataParam = std::make_shared<ngraph::opset3::Parameter>(
+                inDataType, inDataShape);
+        const auto inDataShapeParam = std::make_shared<ngraph::opset3::Parameter>(
+                ngraph::element::i32, ngraph::Shape{inDataShape.size()});
+        inDataShapeParam->set_friendly_name(inDataParam->get_friendly_name() + "/shape");
+
+        m_shapes[inDataShapeParam->get_friendly_name()] = inDataShape;
+        m_parameterVector.push_back(inDataParam);
+        m_parameterVector.push_back(inDataShapeParam);
+
+        const auto dataConstFactor = std::make_shared<ngraph::opset3::Constant>(
+                inDataType, ngraph::Shape{}, 1.5f);
+        const auto shapeConstFactor = std::make_shared<ngraph::opset3::Constant>(
+                ngraph::element::i32, ngraph::Shape{inDataShape.size()}, 1);
+        const auto inDataCopy = std::make_shared<ngraph::opset3::Multiply>(
+                inDataParam, dataConstFactor);
+        const auto inDataShapeCopy = std::make_shared<ngraph::opset3::Multiply>(
+                inDataShapeParam, shapeConstFactor);
+
+        const auto dsr = std::make_shared<ngraph::vpu::op::DynamicShapeResolver>(
+                inDataCopy, inDataShapeCopy);
+
+        return dsr;
+    }
+
+    virtual FunctionWithTestedOp createFunctionWithTestedOp() = 0;
+
+    void SetUp() override {
+        SetRefMode(LayerTestsUtils::RefMode::CONSTANT_FOLDING);
+        configuration[VPU_CONFIG_KEY(DETECT_NETWORK_BATCH)] = CONFIG_VALUE(NO);
+        configuration[VPU_CONFIG_KEY(DISABLE_REORDER)] = CONFIG_VALUE(YES);
+
+        const auto& infFunctionWithTestedOp = createFunctionWithTestedOp();
+        function = infFunctionWithTestedOp.function;
+        const auto testedOp = infFunctionWithTestedOp.testedOp;
+        testedOp->set_output_type(0, testedOp->get_input_element_type(0), ngraph::PartialShape::dynamic(
+                testedOp->get_output_partial_shape(0).rank()));
+        vpu::DynamicToStaticShape().transform(function);
+    }
+
+    InferenceEngine::Blob::Ptr GenerateInput(const InferenceEngine::InputInfo &info) const override {
+        const auto& shapeIt = m_shapes.find(info.name());
+        if (shapeIt == m_shapes.end()) {
+            return LayerTestsCommon::GenerateInput(info);
+        }
+
+        InferenceEngine::Blob::Ptr blob = make_blob_with_precision(info.getTensorDesc());
+        blob->allocate();
+
+        auto dataPtr = InferenceEngine::as<InferenceEngine::MemoryBlob>(blob)->rwmap().as<int32_t*>();
+        for (size_t i = 0; i < blob->size(); ++i) {
+            dataPtr[i] = shapeIt->second[i];
+        }
+
+        return blob;
+    }
+
+    void Validate() override {
+        for (const auto& op : function->get_ordered_ops()) {
+            if (const auto dsr = ngraph::as_type_ptr<ngraph::vpu::op::DynamicShapeResolver>(op)) {
+                dsr->setMode(ngraph::vpu::op::DynamicShapeResolverMode::INFER_DYNAMIC_SHAPE);
```

Comment:
Why do we need these 2 modes?
`INFER_UPPER_BOUND_SHAPE` -> does it mean that we compare OpenVINO actual results vs nGraph upper bound?

Reply:
If DSR was inserted after such operation, then we can argue that this operation had dynamic output. 
With `INFER_UPPER_BOUND_SHAPE` mode DSR just propagates upper-bound shapes, generated in StaticShape* operations.

But in tests, there are no operations with dynamic output and DSR is inserted "artificially" in the begging of Function. So we get a dynamic input.  
Let's say the plugin now copes with the task of processing dynamic input and generate some output data.

Now we have to get a reference for this Function from constant folding process.
But if we still propagate upper-bound shape through DSR, the evaluator sets input HostTensor and output HostTensor with static shapes and we will not able to change them, therefore we will copy garbage data.

To avoid this we need to propagate dynamism through the Function and evaluator sets dynamic Partial shapes to HostTensor. So we can set the real shape to output HostTensor according to the second DSR input and copy only necessary data.

For this, just the second mode `INFER_DYNAMIC_SHAPE` was added. And we compare OpenVINO actual results with nGraph actual results (not upper-bound).

Reply:
Discussed offline, resolving